epochStart/metachain/systemSCs.go

```diff
@@ -327,9 +334,149 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
 		}
 	}
 
+	if s.flagStakingV4Enabled.IsSet() {
+		allNodesKeys := s.getAllNodeKeys(validatorInfos)
+
+		err := s.stakingDataProvider.PrepareStakingData(allNodesKeys)
+		if err != nil {
+			return err
+		}
+
+		err = s.selectNodesFromAuctionList(validatorInfos, randomness)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *systemSCProcessor) selectNodesFromAuctionList(validatorInfoMap map[uint32][]*state.ValidatorInfo, randomness []byte) error {
+	auctionList, numOfValidators := getAuctionListAndNumOfValidators(validatorInfoMap)
+	err := s.sortAuctionList(auctionList, randomness)
+	if err != nil {
+		return err
+	}
+
+	auctionListSize := uint32(len(auctionList))
+	numOfAvailableNodeSlots := core.MinUint32(auctionListSize, s.maxNodes-numOfValidators)
+	s.displayAuctionList(auctionList, numOfAvailableNodeSlots)
+
+	for i := uint32(0); i < numOfAvailableNodeSlots; i++ {
+		auctionList[i].List = string(common.NewList)
+	}
+
 	return nil
 }
 
+func getAuctionListAndNumOfValidators(validatorInfoMap map[uint32][]*state.ValidatorInfo) ([]*state.ValidatorInfo, uint32) {
+	auctionList := make([]*state.ValidatorInfo, 0)
+	numOfValidators := uint32(0)
+
+	for _, validatorsInShard := range validatorInfoMap {
+		for _, validator := range validatorsInShard {
+			if validator.List == string(common.AuctionList) {
+				auctionList = append(auctionList, validator)
+				continue
+			}
+			if isValidator(validator) {
+				numOfValidators++
+			}
+		}
+	}
+
+	return auctionList, numOfValidators
+}
+
+func (s *systemSCProcessor) sortAuctionList(auctionList []*state.ValidatorInfo, randomness []byte) error {
+	validatorTopUpMap, err := s.getValidatorTopUpMap(auctionList)
+	if err != nil {
+		return fmt.Errorf("%w: %v", epochStart.ErrSortAuctionList, err)
+	}
+
+	sort.SliceStable(auctionList, func(i, j int) bool {
+		pubKey1 := auctionList[i].PublicKey
+		pubKey2 := auctionList[j].PublicKey
+
+		nodeTopUpPubKey1 := validatorTopUpMap[string(pubKey1)]
+		nodeTopUpPubKey2 := validatorTopUpMap[string(pubKey2)]
+
+		if nodeTopUpPubKey1.Cmp(nodeTopUpPubKey2) == 0 {
+			return compareByXORWithRandomness(pubKey1, pubKey2, randomness)
+		}
+
+		return nodeTopUpPubKey1.Cmp(nodeTopUpPubKey2) == 1
+	})
+
+	return nil
+}
+
+func (s *systemSCProcessor) getValidatorTopUpMap(validators []*state.ValidatorInfo) (map[string]*big.Int, error) {
+	ret := make(map[string]*big.Int, len(validators))
+
+	for _, validator := range validators {
+		pubKey := validator.PublicKey
+		topUp, err := s.stakingDataProvider.GetNodeStakedTopUp(pubKey)
```

Comment:
discussion is needed here in order to see if we want the following: one entity might register 10 nodes with 3000eGLD, meaning 500 topup per node. but if the topUp limit is at 2500 - that entity might have 5 nodes which enter and 5 which will not. However the first auction system was designed like that - selecting the most possible for an account. So add a task for this - let's discuss and it should be another task if we decide that.

Reply:
Will leave this topic open and create a new Jira task